### PR TITLE
[#22] Add functionality for session

### PR DESCRIPTION
**Pull request by davecaos**
First PR of #22 issue

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,8 +10,9 @@ dep_sumo        = git https://github.com/inaka/sumo_db.git   0.3.5
 dep_cowboy      = git git://github.com/ninenines/cowboy.git  1.0.1
 dep_mixer       = git git://github.com/inaka/mixer.git       0.1.3
 dep_jiffy       = git git://github.com/davisp/jiffy.git      0.11.3
-dep_shotgun     = git git://github.com/inaka/shotgun.git     0.1.8
+dep_shotgun     = git git://github.com/inaka/shotgun.git     0.1.10
 dep_xref_runner = git git://github.com/inaka/xref_runner.git 0.2.2
+dep_uuid 				= git git://github.com/okeuday/uuid.git 		 v1.5.0
```
Comment (elbrujohalcon) [inline Makefile]: too many spaces

File: rel/sys.config [hunk 2]
```diff
@@ -28,6 +28,8 @@
      , {cnf_message, conferl_store_pgsql}
      , {cnf_user   , conferl_store_pgsql}
      , {cnf_vote   , conferl_store_pgsql}
+     , {cnf_session , conferl_store_pgsql}
```
Comment (elbrujohalcon) [inline rel/sys.config]: useless space before ,

File: src/docs/cnf_session.erl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+-spec token(session(), integer()) -> session().
+token(Session, UserId) ->
+  Session#{token => UserId}.
+
+-spec created_at(session()) -> tuple().
+created_at(Session) ->
+  maps:get(created_at, Session).
+
+-spec created_at(session(), tuple()) -> session().
+created_at(Session, CreatedAt) ->
```
Comment (elbrujohalcon) [inline src/docs/cnf_session.erl]: this function will never be needed

File: src/docs/cnf_session.erl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+-spec user_id(session()) -> integer().
+user_id(Session) ->
+  maps:get(user_id, Session).
+
+-spec token(session()) -> integer().
+token(Session) ->
+  maps:get(token, Session).
+
+-spec token(session(), integer()) -> session().
+token(Session, UserId) ->
```
Comment (elbrujohalcon) [inline src/docs/cnf_session.erl]: this function will never be needed, and the parameter is not the UserId, anyway

File: src/docs/cnf_session.erl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+-spec created_at(session(), tuple()) -> session().
+created_at(Session, CreatedAt) ->
+  Session#{created_at => CreatedAt}.
+
+-spec updated_at(session()) -> tuple().
+updated_at(Session) ->
+  maps:get(updated_at, Session).
+
+-spec updated_at(session(), tuple()) -> session().
+updated_at(Session, UpdatedAt) ->
```
Comment (elbrujohalcon) [inline src/docs/cnf_session.erl]: This function will not be needed, sessions can never be updated

File: src/docs/cnf_session.erl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  Session#{updated_at => UpdatedAt}.
+
+-spec is_valid(cnf_session:session()) -> boolean().
+is_valid(Session) ->
```
Comment (elbrujohalcon) [inline src/docs/cnf_session.erl]: This is business logic… it should be on the repo

File: src/docs/cnf_session.erl [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+-export([user_id/1]).
+-export([token/1]).
+-export([token/2]).
+-export([created_at/1]).
+-export([created_at/2]).
+-export([updated_at/1]).
+-export([updated_at/2]).
+-export([is_valid/1]).
+
+-define(MAX_SESSION_DAYS, 42).
```
Comment (elbrujohalcon) [inline src/docs/cnf_session.erl]: This should not be a macro, this should be in the app configuration.

File: src/handlers/cnf_session_handler.erl [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  end.
+
+-spec handle_post(cowboy_req:req(), state()) ->
+  {halt | true, cowboy_req:req(), state()}.
+handle_post(Req, State) ->
+  case cowboy_req:parse_header(<<"authorization">>, Req) of
+    {ok, {<<"basic">>, {Login, _Password}}, _} ->
+      User = cnf_user_repo:find_by_name(Login),
+      Session = cnf_session_repo:register(cnf_user:id(User)),
+      Body = #{token => cnf_session:token(Session)},
```
Comment (elbrujohalcon) [inline src/handlers/cnf_session_handler.erl]: do `cnf_session:to_json` instead

File: src/handlers/cnf_status_handler.erl [hunk 9]
```diff
@@ -32,5 +32,5 @@ handle(Req, State) ->
                     , Req),
   {ok, Req2, State}.
 
-terminate(_Reason, _Req, _State) ->
-  ok.
+%terminate(_Reason, _Req, _State) ->
```
Comment (elbrujohalcon) [inline src/handlers/cnf_status_handler.erl]: don't leave comments in pushed code

File: src/repo/cnf_session_repo.erl [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+-author('David Cao <david.cao@inakanetworks.com>').
+
+-export([register/1]).
+-export([unregister/1]).
+-export([find_by_user/1]).
+-export([find_by_token/1]).
+
+-spec register(non_neg_integer()) -> cnf_session:session().
+register(UserId) ->
+  case find_by_user(UserId) of
```
Comment (elbrujohalcon) [inline src/repo/cnf_session_repo.erl]: Don't do this… let the user create as many sessions as he wants. That way we may track their behaviour better.

File: src/handlers/cnf_session_handler.erl [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+      JsonBody = cnf_session:to_json(Body),
```
Comment (gadgetci) [inline src/handlers/cnf_session_handler.erl]: According to **Dialyzer**:
>  The call cnf_session:to_json(Body:#{}) does not have an opaque term of type cnf_session:session() as 1st argument

File: src/handlers/cnf_session_handler.erl [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      try cnf_user_repo:is_registered(Login, Password) of
+        ok -> {true, Req, Login}
+      catch
+        _Type:Exception ->
+          cnf_utils:handle_exception(Exception, Req, State)
+      end;
+    _ -> {{false, <<"Basic realm=\"conferl\"">>}, Req, State}
+  end.
+
+-spec handle_post(cowboy_req:req(), state()) ->
```
Comment (gadgetci) [inline src/handlers/cnf_session_handler.erl]: According to **Dialyzer**:
>  Invalid type specification for function cnf_session_handler:handle_post/2. The success typing is (cowboy_req:req(),_) -> {{'false',<<_:168>>},cowboy_req:req(),_}

**Pull request merged**